Collaboration service

changjy44 opened the pull request:
**current functionality and flow**
- user2 will create a new room/session upon successful match
- active room/sessions are maintained in a MongoDB collection
  - deleted when either user leaves properly (i.e. clicks the leave button)
- code editor has live updates on both peers' clients
- room/session is closed when either user leaves properly; users are navigated out of session page
- when a user leaves accidentally, they will be prompted to re-join the room/session
  - if their peer closes the room/session, the alert message will say so

closes #32 
closes #33 
closes #34

panpannnnn committed:
Add different language files to editor

panpannnnn committed:
Add different language files to editor

panpannnnn committed:
Update editor to take in files of different extensions

panpannnnn committed:
Update editor to take in files of different extensions

panpannnnn committed:
Merge branch 'collaboration-service' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g43 into collaboration-service

panpannnnn committed:
Remove modal footer for users

panpannnnn committed:
Add search function

panpannnnn committed:
Add split by new line for task description

panpannnnn committed:
change modal scrolling to 'inside'

panpannnnn committed:
add key

panpannnnn committed:
Add scrollable text component

panpannnnn committed:
add code editor to session page

panpannnnn committed:
Allow concurrent running of all services

panpannnnn committed:
more code editor testing

ish1506 committed:
Fix console warning (dup keys in list)

panpannnnn committed:
test split

changjy44 committed:
Create first draft of session service, readded the code block, change socket to emit both uids

changjy44 committed:
Add session creation in frontend

changjy44 committed:
Rename to collaboration service

ish1506 committed:
Merge branch 'collaboration-service' of github.com:CS3219-AY2324S1/ay2324s1-course-assessment-g43 into collaboration-service

ish1506 committed:
Update dev dependencies and npm scripts

ish1506 committed:
Fix client runtime errors

ish1506 commented on client/src/components/Navbar.jsx:
```diff
@@ -472,7 +472,7 @@ const MobileNavItem = ({ label, children, href }) => {
                               redirectToSessionPage(data);
                               toast({
-                                title: `Successfully matched with User #${data.uid} on ${data.complexity} question - ${data.title}`,
+                                title: `Successfully matched with User #${data.firstUserId} on ${data.complexity} question - ${data.title}`,
```
@changjy44 I think the socket messages to the 2 clients will still have to be different, since the clients don't know who's first and second

changjy44 replied:
I disagree. I think we should just define user1 as the first user, and the user2 as the second user. Passing the entire session object for both parties is cleaner in case we want to extend our functionality

ish1506 replied:
Okay, then could you fix the bug on the FE? Coz both users see the same notification message rn

ish1506 committed:
Fix console warning (dup keys in list)

ish1506 committed:
Implement websocket rooms for collab-service

- Also added dynamic route in FE for /session

panpannnnn committed:
remove split

panpannnnn committed:
remove chat area, add boilerplates to code editor

panpannnnn committed:
very minor changes

panpannnnn committed:
Changes to code editor:
-pass in question title
-add default boilerplate code

panpannnnn committed:
more changes

panpannnnn committed:
Add send code, receive code

rachtan27 committed:
code execution backend

changjy44 committed:
Change to y-websocket to faciliate code collaboration

changjy44 committed:
Add more missing files

ish1506 committed:
Set up WSS server for collab-service

ish1506 committed:
Implement language changing syncing in CodeEditor

- WSS server now runs on different port
- Python is set as the default language
- Current bug: code editor initialises without code template

panpannnnn committed:
Create ui for console display and 3 other buttons

changjy44 committed:
Merge branch 'collaboration-service' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g43 into collaboration-service

panpannnnn committed:
Add confirm popups for reset code and change language. disabled buttons when language is whiteboard.

rachtan27 committed:
Merge branch 'collaboration-service' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g43 into code-execution

panpannnnn committed:
Get editor current value for submission

ish1506 committed:
Add question fetching on 'manual' navigation

- Checks ensure that only active sessions can be accessed

changjy44 committed:
Merge branch 'collaboration-service' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g43 into collaboration-service

rachtan27 committed:
Merge branch 'collaboration-service' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g43 into code-execution

panpannnnn committed:
Minor touchup: whiteboard has no default text

panpannnnn committed:
Create resulme session component

panpannnnn committed:
rename whiteboard -> notes

ish1506 committed:
Fix bug in socket change-language event

- Removed roomId due to unknown problem with reading from mobX store

changjy44 committed:
Merge branch 'collaboration-service' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g43 into collaboration-service

changjy44 committed:
Changed schema of session object, changed the leave session endpoint to delete the session, enforce user uniqueness in sessions, improved session creation validation and improved error handling for session creation in frontend

changjy44 committed:
Added syncing of leave rooms between users, fixed a bug where userId was undefined

changjy44 committed:
add missing files

ish1506 committed:
Improve error handling and fix formatting nits

ish1506 committed:
Add ResumeSession conditional rendering

- Added logic to update localStorage when user joins/leaves session
- Added custom Hook to handle conditional rendering of RemoveSession

panpannnnn committed:
set modal to open on default

panpannnnn committed:
change disability setting for run buttons

ish1506 committed:
Add resume session page navigation to popup

ish1506 committed:
Add roomId localStorage cache invalidation logic

panpannnnn committed:
Merge branch 'collaboration-service' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g43 into collaboration-service

panpannnnn committed:
Set to always open

ish1506 committed:
Improve error message (partner closed room case)

rachtan27 committed:
work in progress

rachtan27 committed:
Merge branch 'collaboration-service' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g43 into code-execution

rachtan27 committed:
fix bugs

changjy44 committed:
Merge pull request #38 from CS3219-AY2324S1/code-execution

Code execution

ish1506 committed:
Update npm scripts

ish1506 committed:
Fix typos

panpannnnn committed:
Remove submit button, reformat console drawer

panpannnnn committed:
Change Navbar for normal users, create history page, route to history.

panpannnnn committed:
Create get random question button

panpannnnn committed:
create chat placeholder

changjy44 committed:
Initialise matching service

changjy44 committed:
Finish backend for matching-service

changjy44 committed:
Implement matching rooms in frontend

changjy44 committed:
Enable cancel match on frontend

changjy44 committed:
Add cancel functionality to mobile nav as well

ish1506 committed:
Implement session language persistence

- Language is persisted across a session
- Assumes the peer doesn't change language while user is gone

panpannnnn committed:
Merge branch 'collaboration-service' of https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g43 into collaboration-service

panpannnnn committed:
Update navbar for users

panpannnnn committed:
Change get random question button to icon button

panpannnnn committed:
remove chat modal, to add in future PR

ish1506 committed:
Fix console not opening bug

ish1506 committed:
Merge branch 'collaboration-service' into refactor-matching-service

ish1506 committed:
Refactor to use optional chaining on callbacks

changjy44 committed:
Improved comments and styling

changjy44 committed:
Change matching modal to not be closable for the new navbar

ish1506 committed:
Merge pull request #39 from CS3219-AY2324S1/refactor-matching-service

Refactor matching service
